src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/AppealReferenceNumberHandler.java

```diff
@@ -0,0 +1,69 @@
+package uk.gov.hmcts.reform.iacaseapi.domain.handlers.presubmit;
+
+import static java.util.Arrays.asList;
+import static java.util.Objects.requireNonNull;
+
+import java.util.Optional;
+import org.springframework.stereotype.Service;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.AsylumCase;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.Event;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.callback.Callback;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.callback.PreSubmitCallbackResponse;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.callback.PreSubmitCallbackStage;
+import uk.gov.hmcts.reform.iacaseapi.domain.handlers.PreSubmitCallbackHandler;
+import uk.gov.hmcts.reform.iacaseapi.domain.service.AppealReferenceNumberGenerator;
+
+@Service
+public class AppealReferenceNumberHandler implements PreSubmitCallbackHandler<AsylumCase> {
+
+    private final AppealReferenceNumberGenerator appealReferenceNumberGenerator;
+
+    public AppealReferenceNumberHandler(
+            AppealReferenceNumberGenerator appealReferenceNumberGenerator) {
+
+        this.appealReferenceNumberGenerator = appealReferenceNumberGenerator;
+    }
+
+    @Override
+    public boolean canHandle(PreSubmitCallbackStage callbackStage, Callback<AsylumCase> callback) {
+
+        requireNonNull(callbackStage, "callbackStage must not be null");
+        requireNonNull(callback, "callback must not be null");
+
+        return callbackStage == PreSubmitCallbackStage.ABOUT_TO_SUBMIT
+                && callback.getEvent() == Event.SUBMIT_APPEAL;
+    }
+
+    @Override
+    public PreSubmitCallbackResponse<AsylumCase> handle(
+            PreSubmitCallbackStage callbackStage,
+            Callback<AsylumCase> callback) {
+
+        if (!canHandle(callbackStage, callback)) {
+            throw new IllegalStateException("Cannot handle callback");
+        }
+
+        AsylumCase caseData = callback
```

Comment:
Can we be consistent with the other parts of the code that would call this `AsylumCase` when we know the concrete type?

Reply:
Unsure what you mean for this one, lets discuss when you are in